src/Geopilot.Api/Authorization/GeopilotUserHandler.cs

```diff
@@ -25,24 +28,61 @@ public GeopilotUserHandler(ILogger<GeopilotUserHandler> logger, Context dbContex
     protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context, GeopilotUserRequirement requirement)
     {
         if (context.User.Identity?.IsAuthenticated != true)
+            return;
+
+        var user = await UpdateOrCreateUser(context);
+        if (user is null)
+            return;
+
+        if (requirement.RequireAdmin && !user.IsAdmin)
         {
+            logger.LogWarning("User with id <{UserId}> did not fulfill admin requirement.", user.AuthIdentifier);
             return;
         }
 
-        var dbUser = await dbContext.GetUserByPrincipalAsync(context.User);
-        if (dbUser == null)
+        context.Succeed(requirement);
+    }
+
+    private async Task<User?> UpdateOrCreateUser(AuthorizationHandlerContext context)
+    {
+        var sub = context.User.Claims.FirstOrDefault(claim => claim.Type == JwtRegisteredClaimNames.Sub)?.Value;
+        var email = context.User.Claims.FirstOrDefault(claim => claim.Type == JwtRegisteredClaimNames.Email)?.Value;
+        var name = context.User.Claims.FirstOrDefault(claim => claim.Type == JwtRegisteredClaimNames.Name)?.Value;
+
+        if (sub == null || name == null || email == null)
         {
-            logger.LogWarning("There was a logging attempt for user with id <{UserId}> without corresponding user in database.",
-                context.User.Claims.FirstOrDefault(claim => claim.Type == ContextExtensions.UserIdClaim)?.Value.ReplaceLineEndings(string.Empty));
-            return;
+            logger.LogError("Login failed as not all required claims were provided.");
+            return null;
         }
 
-        if (requirement.RequireAdmin && !dbUser.IsAdmin)
+        var user = await dbContext.Users.FirstOrDefaultAsync(u => u.AuthIdentifier == sub);
+        if (user == null)
         {
-            logger.LogWarning("User with id <{UserId}> did not fulfill admin requirement.", dbUser.AuthIdentifier);
-            return;
+            user = new User { AuthIdentifier = sub, Email = email, FullName = name };
+            await dbContext.Users.AddAsync(user);
+
+            logger.LogInformation("New user (with sub <{Sub}>) has been registered in database.", sub);
+
+            await dbContext.SaveChangesAsync();
+            await ElevateFirstUserToAdmin(user);
+        }
+        else if (user.Email != email || user.FullName != name)
+        {
+            // Update user information in database from provided principal
+            user.Email = email;
+            user.FullName = name;
+            await dbContext.SaveChangesAsync();
         }
 
-        context.Succeed(requirement);
+        return user;
+    }
+
+    private async Task ElevateFirstUserToAdmin(User user)
+    {
+        if (dbContext.Users.Any(u => u.Id == user.Id))
```

Comment:
REQ:
```suggestion
        if (!dbContext.Users.Any(u => u.Id != user.Id))
```

Verhindert gegenüber einem if (dbContext.Users.Any) vor dem erstellen des Users, dass ein Count(*) auf der Datenbank ausgeführt wird.

Reply:
Logik wurde mit [46b8f97](https://github.com/GeoWerkstatt/geopilot/pull/250/commits/46b8f97fc3cced311e35c486aec1806e19a15125) ersetzt.